include/macroni/Translation/Kernel/KernelVisitor.hpp

```diff
@@ -0,0 +1,98 @@
+
+namespace macroni::kernel {
+using rcu_dereference_table =
+    std::unordered_map<const clang::StmtExpr *, const clang::Expr *>;
+
+struct rcu_assign_pointer_parameters {
+  const clang::Expr *p;
+  const clang::Expr *v;
+};
+
+struct rcu_replace_pointer_parameters {
+  const clang::Expr *rcu_ptr;
+  const clang::Expr *ptr;
+  const clang::Expr *c;
+};
+
+using rcu_assign_pointer_table =
+    std::unordered_map<const clang::DoStmt *, rcu_assign_pointer_parameters>;
+
+using rcu_access_pointer_table =
+    std::unordered_map<const clang::StmtExpr *, const clang::Expr *>;
+
+using rcu_replace_pointer_table =
+    std::unordered_map<const clang::StmtExpr *, rcu_replace_pointer_parameters>;
+
+struct kernel_visitor : ::macroni::empty_visitor {
+  [[nodiscard]] explicit kernel_visitor(
+      rcu_dereference_table &rcu_dereference_to_p,
+      rcu_dereference_table &rcu_dereference_bh_to_p,
+      rcu_dereference_table &rcu_dereference_sched_to_p,
+      rcu_assign_pointer_table &rcu_assign_pointer_params,
+      rcu_access_pointer_table &rcu_access_pointer_to_p,
+      rcu_replace_pointer_table &m_rcu_replace_pointer_to_params,
+      vast::acontext_t &actx, vast::mcontext_t &mctx,
+      vast::cg::codegen_builder &bld, vast::cg::meta_generator &mg,
+      vast::cg::symbol_generator &sg, vast::cg::visitor_view view);
+
+  [[nodiscard]] vast::operation visit(const vast::cg::clang_stmt *stmt,
+                                      vast::cg::scope_context &scope) override;
+
+  [[nodiscard]] std::optional<vast::operation>
+  visit_rcu_dereference(const vast::cg::clang_stmt *stmt,
+                        vast::cg::scope_context &scope);
+
+  [[nodiscard]] std::optional<vast::operation>
+  visit_rcu_read_lock_or_unlock(const vast::cg::clang_stmt *stmt,
+                                vast::cg::scope_context &scope);
+
+  [[nodiscard]] std::optional<vast::operation>
+  visit_rcu_assign_pointer(const vast::cg::clang_stmt *stmt,
+                           vast::cg::scope_context &scope);
+
+  [[nodiscard]] std::optional<vast::operation>
+  visit_rcu_access_pointer(const vast::cg::clang_stmt *stmt,
+                           vast::cg::scope_context &scope);
+
+  [[nodiscard]] std::optional<vast::operation>
+  visit_rcu_replace_pointer(const vast::cg::clang_stmt *stmt,
+                            vast::cg::scope_context &scope);
+
+  [[nodiscard]] vast::operation visit(const vast::cg::clang_decl *decl,
+                                      vast::cg::scope_context &scope) override;
+
+  void set_lock_level(mlir::Operation &op);
+
+  void lock_op(mlir::Operation &op);
+
+  void unlock_op(mlir::Operation &op);
+
+  rcu_dereference_table &m_rcu_dereference_to_p;
+  rcu_dereference_table &m_rcu_dereference_bh_to_p;
+  rcu_dereference_table &m_rcu_dereference_sched_to_p;
+  rcu_assign_pointer_table &m_rcu_assign_pointer_params;
+  rcu_access_pointer_table &m_rcu_access_pointer_to_p;
+  rcu_replace_pointer_table &m_rcu_replace_pointer_to_params;
```

Comment:
Wrap this in a POD struct that just carries them along? Much better to see one member here (and in ctor) than these "random" 6. And god forbid you need to modify it.